src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +244,310 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+            MaterializerDeltaResourceSet deltaResourceSetMaterializer = default;
+
+            // We are setting the MaxProtocolVersion to v4.01 when reading bulk update responses.
+            this.RequestInfo.Context.MaxProtocolVersion = ODataProtocolVersion.V401;
+            Exception exception = BaseSaveResult.HandleResponse(this.RequestInfo, (HttpStatusCode)response.StatusCode, response.GetHeader(XmlConstants.HttpODataVersion), () => { return ResponseStream; }, throwOnFailure: false, out Version responseVersion);
+            HeaderCollection headers = new HeaderCollection(response);
+           
+            if (responseStream != null && entityDescriptor.DescriptorKind == DescriptorKind.Entity)
+            {
+                try
+                {
+                    ResponseInfo responseInfo = this.CreateResponseInfo(entityDescriptor);
+                    var responseMessageWrapper = new HttpWebResponseMessage(
+                        headers,
+                        response.StatusCode,
+                        () => this.ResponseStream);
+
+                    ODataMaterializerContext materializerContext = new ODataMaterializerContext(responseInfo, base.MaterializerCache);
+                    deltaResourceSetMaterializer = ODataReaderEntityMaterializer.ParseDeltaResourceSetPayload(responseMessageWrapper, responseInfo, entityDescriptor.Entity.GetType(), materializerContext);
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+
+                    if (!CommonUtil.IsCatchableExceptionType(ex))
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            // Cache the response.
+            this.cachedResponse = new CachedResponse(
+                this.bulkUpdateGraph,
+                headers,
+                (HttpStatusCode)response.StatusCode,
+                responseVersion,
+                deltaResourceSetMaterializer,
+                exception);
+        }
+
+        /// <summary>
+        /// This method traverses the <see cref="BulkUpdateGraph"/> that was used to serialize the bulk update request
+        /// and returns a descriptor for an operation response. When a descriptor gets returned it is marked as visited. 
+        /// </summary>
+        /// <param name="descriptors">The list of descriptors to enumerate.</param>
+        /// <returns>The descriptor for the particular operation being processed.</returns>
+        private Descriptor TraverseBulkUpdateGraph(IEnumerable<Descriptor> descriptors)
+        {
+            foreach (Descriptor item in descriptors)
+            {
+                if (!this.bulkUpdateGraph.Visited(item))
+                {
+                    this.BulkUpdateGraph.AddVisited(item);
+
+                    return item;
+                }
+            }
+
+            return null;
+        }

```

Comment:
Yes this method traverses the passed in list of descriptors and returns the descriptor that has not been visited.  Renaming it could be ok. I could have it in the bulk update graph class.

Reply:
Yea. I think the information is not useful to the `BulkUpdateGraph` and cannot be used outside of this class. Can be moved here.